core_types/src/eav.rs

```diff
@@ -287,87 +357,19 @@ impl EntityAttributeValueStorage for ExampleEntityAttributeValueStorage {
-                    attribute.as_ref(),
-                    &index_query,
-                )
-            })
-            .filter(|e| EntityAttributeValueIndex::filter_on_eav(&e.value(), value.as_ref()))
-            .filter(|e| {
-                index_query
-                    .start()
-                    .map(|start| start <= e.index())
-                    .unwrap_or_else(|| {
-                        let latest = get_latest(e.clone(), map.clone(), index_query.clone())
-                            .unwrap_or_default();
-                        latest.index() == e.index()
-                    })
-            })
-            .filter(|e| {
-                index_query
-                    .end()
-                    .map(|end| end >= e.index())
-                    .unwrap_or_else(|| {
-                        let latest = get_latest(e.clone(), map.clone(), index_query.clone())
-                            .unwrap_or(EntityAttributeValueIndex::default());
-                        latest.index() == e.index()
-                    })
-            })
-            .collect::<BTreeSet<EntityAttributeValueIndex>>();
-
-        Ok(filtered)
+        let lock = self.storage.read()?;
+        let set = (*lock).clone();
+        let iter = set.iter().cloned();
+        Ok(query.run(iter))
     }
 }
 
-pub fn get_latest(
-    eav: EntityAttributeValueIndex,
-    map: BTreeSet<EntityAttributeValueIndex>,
-    index_query: IndexQuery,
-) -> HcResult<EntityAttributeValueIndex> {
-    let filter = map
-        .clone()
-        .into_iter()
-        .filter(|e| EntityAttributeValueIndex::filter_on_eav(&e.entity(), Some(&eav.entity())))
-        .filter(|e| {
-            let prefixes = index_query.prefixes();
-            //find a better way of handling tags with prefixes already
-            let attribute_without_prefix =
-                prefixes.iter().fold(eav.attribute(), |attribute, prefix| {
-                    if eav.attribute().starts_with(prefix) {
-                        let attri = attribute.clone();
-                        attri.replace(prefix, "")
-                    } else {
-                        attribute.clone()
-                    }
-                });
-            EntityAttributeValueIndex::filter_on_eav_with_prefix(
-                &e.attribute(),
-                Some(&attribute_without_prefix),
-                &index_query,
-            )
-        })
-        .filter(|e| EntityAttributeValueIndex::filter_on_eav(&e.value(), Some(&eav.value())));
-    filter.last().ok_or(HolochainError::ErrorGeneric(
-        "Could not get last value".to_string(),
-    ))
-}
-
 impl PartialEq for EntityAttributeValueStorage {
     fn eq(&self, other: &EntityAttributeValueStorage) -> bool {
-        self.fetch_eavi(None, None, None, IndexQuery::default())
-            == other.fetch_eavi(None, None, None, IndexQuery::default())
+        let query = EaviQuery::default();
+        self.fetch_eavi(&query) == other.fetch_eavi(&query)
```

Comment:
I (hopefully) kept this logic the same, which is using the `get_latest` / `LatestByAttribute` method of fetching items, but I wonder if it shouldn't be altered to just return all items? @AshantiMutinta? This is one place where I'm not sure what the original does.

Reply:
Oh so get latest returns the latest record in the set if the dates are not explicitly provided. I think this is fine but if we ever wanted to do a get latest per record we could always modify the way the eavi works but I think from a functionality perspective a get_latest should be just fine